components/cloud/aws/src/s3.rs

```diff
@@ -626,6 +643,95 @@ impl BlobStorage for S3Storage {
     }
 }
 
+struct S3PrefixIter<'cli> {
+    cli: &'cli S3Storage,
+    finished: bool,
+    cont_token: Option<String>,
+    prefix: String,
+}
+
+impl<'cli> S3PrefixIter<'cli> {
+    async fn next_page(&mut self) -> io::Result<Option<Vec<BlobObject>>> {
+        if self.finished {
+            return Ok(None);
+        }
+        let mut input = ListObjectsV2Request::default();
+        input.bucket = String::clone(&self.cli.config.bucket.bucket);
+        input.prefix = Some(self.cli.maybe_prefix_key(&self.prefix));
+        input.continuation_token = self.cont_token.clone();
+        let res = retry_and_count(
+            || self.cli.client.list_objects_v2(input.clone()),
+            "get_one_page",
+        )
+        .await
+        .map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
+        self.finished = !res.is_truncated.ok_or_else(|| {
+            io::Error::new(io::ErrorKind::InvalidData, "no IsTruncated in response")
+        })? || res.next_continuation_token.is_none();
+        self.cont_token = res.next_continuation_token;
+        let data = res
+            .contents
+            .unwrap_or_default()
+            .into_iter()
+            .map(|data| BlobObject {
+                key: self
+                    .cli
+                    .strip_prefix_if_needed(data.key.unwrap_or_default()),
+            })
+            .collect::<Vec<_>>();
+        Ok(Some(data))
+    }
+}
+
+impl DeletableStorage for S3Storage {
+    fn delete(&self, name: &str) -> LocalBoxFuture<'_, io::Result<()>> {
+        let key = self.maybe_prefix_key(name);
+        async move {
+            debug!("delete file from s3 storage"; "key" => %key);
```

Comment:
except log, it's better to record the delete latency in metrics. `CLOUD_REQUEST_HISTOGRAM_VEC` e.g. and also in gcs.